pkg/devfile/adapters/docker/component/adapter.go

```diff
@@ -0,0 +1,219 @@
+package component
+
+import (
+	"fmt"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+
+	"github.com/openshift/odo/pkg/devfile/adapters/common"
+	adaptersCommon "github.com/openshift/odo/pkg/devfile/adapters/common"
+	"github.com/openshift/odo/pkg/devfile/adapters/docker/utils"
+	versionsCommon "github.com/openshift/odo/pkg/devfile/versions/common"
+	"github.com/openshift/odo/pkg/lclient"
+	"github.com/openshift/odo/pkg/log"
+)
+
+// New instantiantes a component adapter
+func New(adapterContext common.AdapterContext, client lclient.Client) Adapter {
+	return Adapter{
+		Client:         client,
+		AdapterContext: adapterContext,
+	}
+}
+
+// Adapter is a component adapter implementation for Kubernetes
+type Adapter struct {
+	Client lclient.Client
+	common.AdapterContext
+}
+
+// Push updates the component if a matching component exists or creates one if it doesn't exist
+func (a Adapter) Push(parameters common.PushParameters) (err error) {
+	componentExists := utils.ComponentExists(a.Client, a.ComponentName)
+
+	if componentExists {
+		err = a.updateComponent()
+	} else {
+		err = a.createComponent()
+	}
+
+	if err != nil {
+		return errors.Wrap(err, "unable to create or update component")
+	}
+
+	return nil
+}
+
+// DoesComponentExist returns true if a component with the specified name exists, false otherwise
+func (a Adapter) DoesComponentExist(cmpName string) bool {
+	return utils.ComponentExists(a.Client, cmpName)
+}
+
+func (a Adapter) createComponent() (err error) {
+	componentName := a.ComponentName
+
+	volumeLabels := map[string]string{
+		"component": componentName,
+		"type":      "projects",
+	}
+
+	supportedComponents := adaptersCommon.GetSupportedComponents(a.Devfile.Data)
+	if len(supportedComponents) == 0 {
+		return fmt.Errorf("No valid components found in the devfile")
+	}
+
+	// Create a docker volume to store the project source code
+	volume, err := a.Client.CreateVolume(volumeLabels)
+	if err != nil {
+		return errors.Wrapf(err, "Unable to create project source volume for component %s", componentName)
+	}
+
+	projectVolumeName := volume.Name
+
+	// Loop over each component and start a container for it
+	for _, comp := range supportedComponents {
+		s := log.Spinner("Pulling image " + *comp.Image)
+		defer s.End(false)
+
+		// Pull the image (as the Docker daemon requires it to be on the system before starting it)
+		err = a.Client.PullImage(*comp.Image)
+		if err != nil {
+			return errors.Wrapf(err, "Unable to pull %s image", *comp.Image)
+		}
+		s.End(true)
+
+		// Start the container
+		err = a.startContainer(componentName, projectVolumeName, comp)
+		if err != nil {
+			return errors.Wrapf(err, "Unable to start container for devfile component %s", *comp.Alias)
+		}
+		glog.V(3).Infof("Successfully created container %s for component %s", *comp.Image, componentName)
+	}
```

Comment:
Can this be moved to a func in docker/utils/utils.go to be consistent with kubernetes package, which does all the heavy handling like kubernetes component? Same for the update component function..

Reply:
Yeah, makes sense to me.

Reply:
Actually, this should stay in adapter.go since it relies on the Adapter struct defined in this file (it's also equivalent to code in createOrUpdateComponent in kubernetes/component/adapter.go)

Reply:
Ended up moving these functions into their own file (utils.go) in the `component` package, like what was done for the storage adapter.